src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -3613,10 +3611,10 @@ public Type visitTypeVar(TypeVar t, Void ignored) {
                     return to.head.withTypeVar(t);
                 }
                 if (allowUniversalTVars &&
-                        t.createdFromUniversalTypeVar &&
+                        !t.isValueProjection() &&
                         from.head.hasTag(TYPEVAR) &&
-                        ((TypeVar)from.head).referenceProjection != null &&
-                        t.equalsIgnoreMetadata(((TypeVar)from.head).referenceProjection)) {
+                        ((TypeVar)from.head).projection != null &&
```

Comment:
Shouldn't this test be `isUniversal()` ?

Reply:
doing this change right now breaks the build, I think it is because we generate reference projections for universal type vars in a lazy way, so some universal type variables could have its projection field set to null

Reply:
ugh - I understand - moving forward it'd be nice not to depend on tricky init semantics.

Reply:
yep, will fix this